test/ooni/dns_injection.cpp

```diff
@@ -5,11 +5,89 @@
 #define CATCH_CONFIG_MAIN
 #include "src/ext/Catch/single_include/catch.hpp"
 
+#include "src/ooni/collector_client_impl.hpp"
 #include "src/ooni/dns_injection_impl.hpp"
 #include <measurement_kit/common.hpp>
 
-using namespace mk;
 using namespace mk::ooni;
+using namespace mk::report;
+using namespace mk;
+
+TEST_CASE("Check dns-injection output for injected entry") {
+    loop_with_initial_event([=]() {
+        dns_injection("8.8.8.8:53", "nexa.polito.it", [=](Error err,
+                                                          Var<Entry> entry) {
+            // We have called the DNS injection test with an actually
+            // valid DNS server, therefore it should return no error and
+            // this should be interpreted as DNS injection.
+            REQUIRE(!err);
+            REQUIRE(entry->is_object());
+            bool injected = (*entry)["injected"];
+            REQUIRE(injected == true);
+            Entry queries = (*entry)["queries"];
+            REQUIRE(queries.is_array());
+            REQUIRE(queries.size() == 1);
+            Entry first_query = queries[0];
+            Entry failure = first_query["failure"];
+            REQUIRE(failure.is_null());
+            std::string hostname = first_query["hostname"];
+            REQUIRE(hostname == "nexa.polito.it");
+            std::string query_type = first_query["query_type"];
+            REQUIRE(query_type == "A");
+            std::string resolver_hostname = first_query["resolver_hostname"];
+            REQUIRE(resolver_hostname == "8.8.8.8");
+            int resolver_port = first_query["resolver_port"];
+            REQUIRE(resolver_port == 53);
+            Entry answers = first_query["answers"];
+            REQUIRE(answers.is_array());
+            REQUIRE(answers.size() == 1);
+            Entry first_answer = answers[0];
+            std::string answer_type = first_answer["answer_type"];
+            REQUIRE(answer_type == "A");
+            std::string ipv4 = first_answer["ipv4"];
+            REQUIRE(ipv4 == "130.192.16.172");
+            int ttl = first_answer["ttl"];
+            REQUIRE(ttl >= 0);
```

Comment:
The above is quite ugly, but I doubt that it would be much better if using `nlohmann::json`. It could perhaps be possible, using `nlohmann::json` to define an expected object to use as template to verify the received one.